src/heimdall/worker.cc

```diff
@@ -0,0 +1,903 @@
+                                reverse_traffic, edge_info);
+
+      // Collect the end node of this edge for the nodes layer
+      unique_nodes.insert(edge->endnode());
+    }
+  }
+
+  return unique_nodes;
+}
+
+void heimdall_worker_t::build_nodes_layer(vtzero::tile_builder& tile,
+                                          const std::unordered_set<baldr::GraphId>& unique_nodes,
+                                          const TileProjection& projection) {
+  // Create nodes layer builder
+  NodesLayerBuilder nodes_builder(tile);
+
+  // Render unique nodes as points
+  for (const auto& node_id : unique_nodes) {
+    baldr::graph_tile_ptr node_tile;
+    const auto* node = reader_->nodeinfo(node_id, node_tile);
+    if (!node || !node_tile) {
+      continue;
+    }
+
+    // Get node position
+    auto node_ll = node->latlng(node_tile->header()->base_ll());
+
+    // Convert to Web Mercator and then to tile coordinates
+    double merc_x = lon_to_merc_x(node_ll.lng());
+    double merc_y = lat_to_merc_y(node_ll.lat());
+
+    double norm_x = (merc_x - projection.tile_merc_minx) / projection.tile_merc_width;
+    double norm_y = (projection.tile_merc_maxy - merc_y) / projection.tile_merc_height;
+
+    int32_t tile_x = static_cast<int32_t>(std::round(norm_x * projection.tile_extent));
+    int32_t tile_y = static_cast<int32_t>(std::round(norm_y * projection.tile_extent));
+
+    // Only render nodes that are within the tile (including buffer)
+    if (tile_x < -projection.tile_buffer ||
+        tile_x > projection.tile_extent + projection.tile_buffer ||
+        tile_y < -projection.tile_buffer ||
+        tile_y > projection.tile_extent + projection.tile_buffer) {
+      continue;
+    }
+
+    // Add feature using the builder
+    nodes_builder.add_feature(vtzero::point{tile_x, tile_y}, node_id, node);
+  }
+}
+
+std::string
+heimdall_worker_t::render_tile(uint32_t z, uint32_t x, uint32_t y, bool return_shortcuts) {
+  // Validate tile coordinates
+  uint32_t max_coord = (1u << z);
+  if (x >= max_coord || y >= max_coord || z > 30) {
+    throw valhalla_exception_t{400, "Invalid tile coordinates"};
+  }
+
+  // Calculate tile bounding box
+  auto bounds = tile_to_bbox(z, x, y);
+
+  // Create vector tile
+  vtzero::tile_builder tile;
+
+  // Don't render anything below minimum zoom level
+  if (z < min_zoom_) {
+    return tile.serialize();
+  }
+
+  // Query edges within the tile bounding box
+  auto edge_ids = candidate_query_.RangeQuery(bounds);
+
+  // Pre-compute Web Mercator projection tile bounds (once for all edges)
+  const int32_t TILE_EXTENT = 4096;
+  const int32_t TILE_BUFFER = 128;
+
+  TileProjection projection{lon_to_merc_x(bounds.minx()),
+                            lon_to_merc_x(bounds.maxx()),
+                            lat_to_merc_y(bounds.miny()),
+                            lat_to_merc_y(bounds.maxy()),
+                            lon_to_merc_x(bounds.maxx()) - lon_to_merc_x(bounds.minx()),
```

Comment:
these 2 seem unnecessary to pass as args? the TileProjection ctor could compute it without lon_to_merc_y etc